### ISSUE TITLE
Poor single class detector 

### ISSUE THREAD

**Issue by gan3sh500** [david8862/keras-YOLOv3-model-set#82]
I was able to get pretty good accuracies training on PascalVOC. But if I try to train on a single class like PascalVOC dogs or custom data with single class, the predictions are very low confidence and confidence loss doesn’t drop enough to bring them up. Is this something you’ve seen or know why it would happen?

**Comment by mvanlierBCG** [comment]
@gan3sh500 I am facing the same issue and would be very interested in how to fix this!

**Comment by david8862** [comment]
@gan3sh500  @mvanlierBCG  what's your model type and train config cmd?

**Comment by gan3sh500** [comment]
For me V4 doesn't get anywhere near darknet mAP. Especially for small objects. I am using your defaults except 1e-4 lr and SGD nestrov 0.94 momentum.

**Comment by pingchesu** [comment]
@gan3sh500 I got good result at license plate detection(One class detection) with yolov4 darknet. Have you tried put lots of negative samples(No bounding box and label) during training? It's gave me really good result!!

**Comment by gan3sh500** [comment]
I don't have much negatives to use in my dataset. I'll try adding. @david8862 Did you compare with Alexey's darknet or the ultralytics yolov3/5 repos. They train way faster and higher mAP.

**Comment by david8862** [comment]
@gan3sh500 I didn't make a peer-to-peer comparison with Alexey's official darknet or ultralytics' pytorch revision. The main purpose of this repo is trying to build an easy & quick end-to-end on-device deploy pipeline with YOLO detection model family, especially for those light-weight models. For the standard YOLOv3/v4 model structure with darknet based backbone, keras implementation here seems do have a bad performance.

**Comment by gan3sh500** [comment]
@david8862 I am doing comparisons and found difference in how you match anchors and how ultralytics and darknet YOLOv4 does. This repo does gt box matched with highest iou anchor. I believe this was the standard approach till YOLOv3. YOLOv4 in bag of freebies has "IT: IoU threshold - using multiple anchors for a single ground truth IoU (truth, anchor) > IoU threshold". This is present in ultralytics code but not in this repo.

**Comment by david8862** [comment]
> @david8862 I am doing comparisons and found difference in how you match anchors and how ultralytics and darknet YOLOv4 does. This repo does gt box matched with highest iou anchor. I believe this was the standard approach till YOLOv3. YOLOv4 in bag of freebies has "IT: IoU threshold - using multiple anchors for a single ground truth IoU (truth, anchor) > IoU threshold". This is present in ultralytics code but not in this repo.

@gan3sh500 many thanks for the clue. Yes that's one of the mentioned freebies in YOLOv4 paper. Could you help to point out the implementations? I'm now working on other projects and may not have much time to dig into their code

**Comment by gan3sh500** [comment]
See around line 420-450 https://github.com/ultralytics/yolov3/blob/master/utils/utils.py#L420

**Comment by david8862** [comment]
> See around line 420-450 https://github.com/ultralytics/yolov3/blob/master/utils/utils.py#L420. Please use it as inspiration so as to keep your code MIT license. 😄

Sure and thanks again. I'll check and try to update this repo, but may not in time due to bandwidth limit...

**Comment by gan3sh500** [comment]
I am trying to fix these myself. Hopefully I can do and PR.

**Comment by david8862** [comment]
> I am trying to fix these myself. Hopefully I can do and PR.

Great thanks! Pls feel free to let me know if you need any help

**Comment by frantz03** [comment]
> @gan3sh500 I got good result at license plate detection(One class detection) with yolov4 darknet. Have you tried put lots of negative samples(No bounding box and label) during training? It's gave me really good result!!

Hey, can you maybe specify how you did the training. I try to do something similar, but is has not succeeded so far :)

**Comment by mvanlierBCG** [comment]
@gan3sh500 @pingchesu Could anyone point me out why using more _negative samples(No bounding box and label) could lead to better single class detection performance? Any paper references would be nice.

**Comment by pingchesu** [comment]
@mvanlierBCG  Refer to https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

desirable that your training dataset include images with non-labeled objects that you do not want to detect - negative samples without bounded box (empty .txt files) - use as many images of negative samples as there are images with objects

There are many enhance tips, maybe you can try which tip can improve your performance.

By the way, I use openimage this tool to collect my negative samples.
